nixos/modules/services/misc/paperless.nix

```diff
@@ -0,0 +1,194 @@
+      default = "localhost";
+      description = "Server listening address.";
+    };
+
+    port = mkOption {
+      type = types.int;
+      default = 28981;
+      description = "Server port to listen on.";
+    };
+
+    extraConfig = mkOption {
+      type = types.attrs;
+      default = {};
+      description = ''
+        Extra paperless config options.
+
+        The config values are evaluated as double-quoted Bash string literals.
+
+        See <literal>paperless-src/paperless.conf.example</literal> for available options.
+
+        To enable user authentication, set <literal>PAPERLESS_DISABLE_LOGIN = "false"</literal>
+        and run the shell command <literal>$dataDir/paperless-manage createsuperuser</literal>.
+
+        To define secret options without storing them in /nix/store, use the following pattern:
+        <literal>PAPERLESS_PASSPHRASE = "$(&lt; /etc/my_passphrase_file)"</literal>
+      '';
+      example = literalExample ''
+        {
+          PAPERLESS_OCR_LANGUAGE = "deu";
+        }
+      '';
+    };
+
+    user = mkOption {
+      type = types.str;
+      default = defaultUser;
+      description = "User under which Paperless runs.";
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.paperless;
+      defaultText = "pkgs.paperless";
+      description = "The Paperless package to use.";
+    };
+
+    manage = mkOption {
+      type = types.package;
+      readOnly = true;
+      default = manage;
+      description = ''
+        A script to manage the Paperless instance.
+        It wraps Django's manage.py and is also available at
+        <literal>$dataDir/manage-paperless</literal>
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable (
+    let
+      preStart = ''
+        if [[ ! -e "${cfg.dataDir}" ]]; then
+          install -o ${cfg.user} -g $(id -gn ${cfg.user}) -d "${cfg.dataDir}"
+        fi
+        if [[ $(readlink ${cfg.dataDir}/paperless-manage) != ${manage} ]]; then
+          ln -sf ${manage} ${cfg.dataDir}/paperless-manage
+        fi
+      '' + optionalString cfg.consumptionDirIsPublic ''
+        if [[ ! -e "${cfg.consumptionDir}" ]]; then
+          install -o ${cfg.user} -g $(id -gn ${cfg.user}) -m 777 -d "${cfg.consumptionDir}"
+        fi
+      '' + ''
+        exec ${pkgs.libuuid}/bin/runuser -u "${cfg.user}" -- ${setupDB}
+      '';
+
+      # Auto-migrate on first run or if the package has changed
+      setupDB = pkgs.writers.writeBash "setupDB" ''
+        set -e
+        ${manage.setupEnv}
+        versionFile="$PAPERLESS_DBDIR/src-version"
+
+        if [[ $(cat "$versionFile" 2>/dev/null) != ${cfg.package} ]]; then
+          python $paperlessSrc/manage.py migrate
+          echo ${cfg.package} > "$versionFile"
+        fi
+      '';
+    in
+      {
+        systemd.services.paperless-consumer = {
+          description = "Paperless document consumer";
+          serviceConfig = {
+            PermissionsStartOnly = true;
```

Comment:
In the future, this should be replaced by tmpfiles.d: https://github.com/NixOS/nixpkgs/pull/56265. But let's leave that for a future PR.

Reply:
`tmpfiles.d` really simplifies the implementation. Fixed here: https://github.com/NixOS/nixpkgs/pull/44086/commits/b5391a0db58a73b10f85e3bd32d73943122e9982